Add Poetry and Python 3.8.5

gkapfham opened the pull request:
This pull request adds support for the following programs to this Docker container:

- Support for Python 3.8.5 through the use of the Debian Buster base image that provides Python 3.8.5
- Support for the Poetry tool for managing Python virtual environments and application dependencies
- The `git` command-line client, which will allow for the use of, for instance, `git status` in the Docker container

The reasons for installing these two programs are as follows:

- Several courses are now using Python 3.8. However, students who are using Windows are having trouble installing this program through the Pyenv-Win tool. This means that these students can now use the Docker container instead of struggling to install Python 3.8.5 on their laptops.

- Instead of using Pipenv for managing virtual environments and packages, I am now starting to use Poetry as it is faster and has more features. Including this program in the Docker container will therefore support future courses as well.

It is important to note that this Docker container no longer uses Alpine, instead opting for Debian Buster. I have tested almost ten different assignments with this Docker container and found that it always worked correctly, focusing on assignments from the following courses: CMPSC 100, CMPSC 101, CMPSC 102, CMPSC 203, and CMPSC 302.

There are several important points to consider about using this Docker container instead of the previous one:

- Since it uses Python 3.8.5 students who previously used the old Docker container will find that their virtual environments in the `.dockagator` directory no longer work correctly. The only way around this issue is to have the delete the two (or three) directories inside of `.dockagator`. In all of my testing, I can confirm that deleting these directories always allowed `gradle grade` to pass when it should.

- Since this Docker container is based on Debian Buster instead of Alpine, this means that the diagnostic display from `gradle grade` and the prompt in the terminal window are slightly different. For instance, here is what the prompt now looks like:

```
root@759d4cbac1bc:/project#
```

this means that some students and student technical leaders may be confused about whether or not  they are actually in the Docker container.

- This Docker container is larger than the one before by about 500 MB. During my testing, I did not notice a significant difference is execution time with this Docker container, once it was downloaded. However, it is possible that using this Docker container will lead to longer initial download times, especially for those students who are using slower laptops.

gkapfham committed:
Add starting ENV setting for the Python section to the Dockerfile

gkapfham committed:
Add poetry directories to the PATH environment variable in Dockerfile

gkapfham committed:
Improve the comments in the environment variables for Python.

gkapfham committed:
Add a temporary version of Dockerfile that does not parse correctly

gkapfham committed:
Add all of the required packages for building Python through pyenv.

gkapfham committed:
Improve the comments in the Dockerfile and remove a duplicate line

gkapfham committed:
Remove the use of proselint from Dockerfile and improve comments

gkapfham committed:
Remove proselint from the Dockerfile

gkapfham committed:
Add a comment in the Dockerfile for the VOLUME line

gkapfham committed:
Remove the use of apk add to install python3 as it is out of date

gkapfham committed:
Add python3 back to the apk add line in the Dockerfile

Michionlion commented:
A possible solution to the space issue with these updates, and the issues with double installations, would be to use the `python:3.8.5-alpine` image as a base. I'll do some testing with that to see if it would work.

gkapfham committed:
Avoid linting errors by changing way that Poetry installs in Dockerfile

gkapfham committed:
Improve the comments and the echo output in the Dockerfile

gkapfham committed:
Move some contents in the Dockerfile to the new approach

gkapfham committed:
Install packages with apt-get and improve diagnostics in Dockerfile

gkapfham committed:
Improve a diagnostic message in the Dockerfile

gkapfham committed:
Add comments about the base image to the Dockerfile

Michionlion commented on Dockerfile:
```diff
@@ -1,31 +1,84 @@
 ENV GATORGRADER_DIR=/root/.local/share/
 
+# Set Python version
+ARG PYTHON_VERSION='3.8.5'
```
Where is this referenced? Does a program end up using or is it just for metadata purposes?

gkapfham replied:
This variable should be delete. I will do so after merging in your new content that installs `sdkman`.

Michionlion commented on Dockerfile:
```diff
@@ -1,31 +1,84 @@
+    # Specify the paths for using requirements and virtual environments;
+    # this is where the requirements + virtual environment will live
+    PYSETUP_PATH="/opt/pysetup" \
+    VENV_PATH="/opt/pysetup/.venv"
```
Is this for project venvs or just poetry's? If it's for projects, we probably want to stick them in `/root/.local/share` so that they persist over container runs (or not if that was designed).

gkapfham replied:
I was initially thinking not to persist them in case a student wants to switch between running Programs in the container versus running then regularly on their laptop. What do you think is the best thing to do in this case?

Michionlion replied:
Will it cause a greater setup time for the container every run, or will it be fairly minimal? If it's a long time (10 seconds or more would be the max I think), then we should probably persist it.

gkapfham replied:
I will go ahead and move this into the /root/.local/share directory. Would that be okay?

Michionlion replied:
That sounds good to me.

gkapfham replied:
Hi @Michionlion Saejin, I moved it to the suggested directory. With that said, I am not noticing that the Docker container stores anything inside of the specified directory, suggesting that perhaps the variable is not needed. With that said, I am currently inclined to keep the variables in the Dockerfile because I don't notice any adverse affects and because they are mentioned in the GitHub issue tracker as being a part of a "representative Dockerfile".

Michionlion committed:
Update Dockerfile for better compatibility

This includes:
- Installing more recent Gradle
- Running setup commands in a bash login shell
- Pinning the Poetry version
- a few other minor tweaks

gkapfham committed:
Move the PYSETUP_PATH and VENV_PATH to the /root/.local/share

gkapfham commented:
Hello @Michionlion I have finished making changes to the code in this pull request. If you have time can you please merge it? Once you merge, I can announce the new Docker container to the students in my classes and to the faculty. Thanks with your expert help on this PR, it is greatly appreciated.
